File: go/processor/extractor/datastore.go
// @docspec:module {
//   id: "docspec-go-extractor-datastore",
//   name: "Data Store Extractor",
//   description: "Detects database, cache, and message broker usage by analyzing import paths for GORM, database/sql, Redis, Kafka, NATS, RabbitMQ, and MongoDB, then extracts table names and topic references.",
//   since: "3.0.0"
// }
package extractor

import (
	"go/ast"
	"strings"
)

// DataStoreExtractor detects data store usage in Go source code:
// @docspec:boundary "classpath-safe extraction"
//   - GORM models (gorm.Model embedding, gorm struct tags)
//   - database/sql usage (sql.DB, sql.Open)
//   - Redis clients (go-redis, redigo)
//   - Message brokers (sarama/Kafka, NATS, RabbitMQ)
//   - MongoDB (mongo-driver)
//   - Comment-based docspec tags (//docspec:datastore, //docspec:table)
//
// This is the Go equivalent of the Java DataStoreExtractor which detects
// JPA entities, Spring Data repositories, Redis, and Kafka usage.
type DataStoreExtractor struct{}

// Known import paths for data store libraries.
var dataStoreImportMap = map[string]struct {
	storeID   string
	storeName string
	storeType string
}{
	"gorm.io/gorm":                             {"rdbms", "Primary Database", "rdbms"},
	"gorm.io/driver/postgres":                  {"rdbms", "PostgreSQL", "rdbms"},
	"gorm.io/driver/mysql":                     {"rdbms", "MySQL", "rdbms"},
	"gorm.io/driver/sqlite":                    {"rdbms", "SQLite", "rdbms"},
	"gorm.io/driver/sqlserver":                 {"rdbms", "SQL Server", "rdbms"},
	"database/sql":                             {"rdbms", "Primary Database", "rdbms"},
	"github.com/jmoiron/sqlx":                  {"rdbms", "Primary Database", "rdbms"},
	"github.com/jackc/pgx/v4":                  {"rdbms", "PostgreSQL", "rdbms"},
	"github.com/jackc/pgx/v5":                  {"rdbms", "PostgreSQL", "rdbms"},
	"github.com/go-sql-driver/mysql":           {"rdbms", "MySQL", "rdbms"},
	"github.com/redis/go-redis/v9":             {"redis", "Redis", "cache"},
	"github.com/go-redis/redis/v8":             {"redis", "Redis", "cache"},
	"github.com/go-redis/redis/v9":             {"redis", "Redis", "cache"},
	"github.com/gomodule/redigo/redis":         {"redis", "Redis", "cache"},
	"github.com/IBM/sarama":                    {"kafka", "Kafka", "message-broker"},
	"github.com/Shopify/sarama":                {"kafka", "Kafka", "message-broker"},
	"github.com/segmentio/kafka-go":            {"kafka", "Kafka", "message-broker"},
	"github.com/confluentinc/confluent-kafka-go/kafka": {"kafka", "Kafka", "message-broker"},
	"github.com/nats-io/nats.go":               {"nats", "NATS", "message-broker"},
	"github.com/streadway/amqp":                {"rabbitmq", "RabbitMQ", "message-broker"},
	"github.com/rabbitmq/amqp091-go":           {"rabbitmq", "RabbitMQ", "message-broker"},
	"go.mongodb.org/mongo-driver/mongo":        {"mongodb", "MongoDB", "document-store"},
}

// @docspec:deterministic
func (e *DataStoreExtractor) Name() string {
	return "data-store"
}

// IsAvailable reports true if any file imports a known data store package.
//
// @docspec:deterministic
func (e *DataStoreExtractor) IsAvailable(ctx *ProcessorContext) bool {
	for imp := range dataStoreImportMap {
		if hasImport(ctx, imp) {
			return true
		}
	}
	return false
}

// Extract scans files for data store patterns and populates ctx.DataStores.
//
// @docspec:method { since: "3.0.0" }
// @docspec:intentional "Three-pass extraction: detect stores from imports, extract GORM table names, then extract Kafka topics"
func (e *DataStoreExtractor) Extract(ctx *ProcessorContext) {
	detectedStores := make(map[string]bool)

	// First pass: detect stores from imports
	for _, file := range ctx.Files {
		for _, imp := range file.Imports {
			path := strings.Trim(imp.Path.Value, `"`)
			if info, ok := dataStoreImportMap[path]; ok {
				if !detectedStores[info.storeID] {
					detectedStores[info.storeID] = true
					addDataStoreIfAbsent(ctx, info.storeID, info.storeName, info.storeType)
				}
			}
		}
	}

	// Second pass: extract tables from GORM models
	for _, file := range ctx.Files {
		imports := fileImports(file)
		hasGorm := imports["gorm.io/gorm"]

		for _, decl := range file.Decls {
			gd, ok := decl.(*ast.GenDecl)
			if !ok {
				continue
			}
			for _, spec := range gd.Specs {
				ts, ok := spec.(*ast.TypeSpec)
				if !ok {
					continue
				}
				st, ok := ts.Type.(*ast.StructType)
				if !ok || st.Fields == nil {
					continue
				}

				// Check for docspec:table comment tag
				tableName := ""
				if gd.Doc != nil {
					for _, c := range gd.Doc.List {
						text := strings.TrimSpace(strings.TrimPrefix(c.Text, "//"))
						if strings.HasPrefix(text, "docspec:table") {
							parts := strings.SplitN(text, " ", 2)
							if len(parts) == 2 {
								tableName = strings.TrimSpace(parts[1])
							}
						}
						if strings.HasPrefix(text, "docspec:datastore") {
							parts := strings.Fields(text)
							if len(parts) >= 4 {
								id := parts[1]
								name := parts[2]
								stype := parts[3]
								addDataStoreIfAbsent(ctx, id, name, stype)
							}
						}
					}
				}

				// Detect GORM models by checking for gorm.Model embedding or gorm struct tags
				isGormModel := false
				if hasGorm {
					for _, field := range st.Fields.List {
						// Check for anonymous gorm.Model field
						if len(field.Names) == 0 {
							if sel, ok := field.Type.(*ast.SelectorExpr); ok {
								if ident, ok := sel.X.(*ast.Ident); ok {
									if ident.Name == "gorm" && sel.Sel.Name == "Model" {
										isGormModel = true
										break
									}
								}
							}
						}
						// Check for gorm struct tags
						if field.Tag != nil {
							tag := field.Tag.Value
							if strings.Contains(tag, `gorm:"`) {
								isGormModel = true
								break
							}
						}
					}
				}

				if isGormModel || tableName != "" {
					if tableName == "" {
						tableName = inferTableName(ts.Name.Name)
					}
					store := findOrAddStore(ctx, "rdbms", "Primary Database", "rdbms")
					addTableIfAbsent(store, tableName)
				}
			}
		}
	}

	// Third pass: extract Kafka topics
	for _, file := range ctx.Files {
		imports := fileImports(file)
		hasKafka := false
		for imp := range imports {
			if strings.Contains(imp, "sarama") || strings.Contains(imp, "kafka") {
				hasKafka = true
				break
			}
		}

		if !hasKafka {
			continue
		}

		for _, decl := range file.Decls {
			fn, ok := decl.(*ast.FuncDecl)
			if !ok || fn.Body == nil {
				continue
			}

			// Look for topic string literals in kafka-related calls
			ast.Inspect(fn.Body, func(n ast.Node) bool {
				call, ok := n.(*ast.CallExpr)
				if !ok {
					return true
				}

				sel, ok := call.Fun.(*ast.SelectorExpr)
				if !ok {
					return true
				}

				// Common patterns: producer.SendMessage, consumer.ConsumePartition, etc.
				methodName := strings.ToLower(sel.Sel.Name)
				if strings.Contains(methodName, "topic") || strings.Contains(methodName, "consume") ||
					strings.Contains(methodName, "produce") || strings.Contains(methodName, "send") ||
					strings.Contains(methodName, "subscribe") {
					for _, arg := range call.Args {
						if lit, ok := arg.(*ast.BasicLit); ok {
							topic := strings.Trim(lit.Value, `"`)
							if topic != "" && !strings.Contains(topic, " ") {
								store := findOrAddStore(ctx, "kafka", "Kafka", "message-broker")
								addTopicIfAbsent(store, topic)
							}
						}
					}
				}

				return true
			})
		}
	}
}

// addDataStoreIfAbsent adds a data store to the context if not already present.
//
// @docspec:preserves "no duplicate data store entries by ID"
func addDataStoreIfAbsent(ctx *ProcessorContext, id, name, storeType string) {
	for _, ds := range ctx.DataStores {
		if ds.ID == id {
			return
		}
	}
	ctx.DataStores = append(ctx.DataStores, DataStore{
		ID:   id,
		Name: name,
		Type: storeType,
	})
}

// findOrAddStore finds an existing store by ID or adds a new one.
//
// @docspec:preserves "no duplicate data store entries by ID"
func findOrAddStore(ctx *ProcessorContext, id, name, storeType string) *DataStore {
	for i := range ctx.DataStores {
		if ctx.DataStores[i].ID == id {
			return &ctx.DataStores[i]
		}
	}
	ctx.DataStores = append(ctx.DataStores, DataStore{
		ID:   id,
		Name: name,
		Type: storeType,
	})
	return &ctx.DataStores[len(ctx.DataStores)-1]
}

// addTableIfAbsent adds a table name to a store if not already present.
//
// @docspec:preserves "no duplicate table names within a data store"
func addTableIfAbsent(store *DataStore, tableName string) {
	for _, t := range store.Tables {
		if t == tableName {
			return
		}
	}
	store.Tables = append(store.Tables, tableName)
}

// addTopicIfAbsent adds a topic to a store if not already present.
//
// @docspec:preserves "no duplicate topic names within a data store"
func addTopicIfAbsent(store *DataStore, topicName string) {
	for _, t := range store.Topics {
		if t.Name == topicName {
			return
		}
	}
	store.Topics = append(store.Topics, DataStoreTopic{Name: topicName})
}

// inferTableName converts a Go struct name to a table name using snake_case pluralization.
// Example: UserProfile -> user_profiles, Order -> orders
//
// @docspec:deterministic
func inferTableName(name string) string {
	var result strings.Builder
	for i, r := range name {
		if i > 0 && r >= 'A' && r <= 'Z' {
			result.WriteByte('_')
		}
		result.WriteRune(r)
	}
	snake := strings.ToLower(result.String())

	// Simple pluralization
	if strings.HasSuffix(snake, "s") || strings.HasSuffix(snake, "x") ||
		strings.HasSuffix(snake, "ch") || strings.HasSuffix(snake, "sh") {
		return snake + "es"
	}
	if strings.HasSuffix(snake, "y") && len(snake) > 1 {
		prev := snake[len(snake)-2]
		if prev != 'a' && prev != 'e' && prev != 'i' && prev != 'o' && prev != 'u' {
			return snake[:len(snake)-1] + "ies"
		}
	}
	return snake + "s"
}

File: rust/processor/src/extractor/datastore.rs
//! Detect data store usage in Rust projects.
//!
//! @docspec:module {
//!   id: "docspec-rust-extractor-datastore",
//!   name: "Data Store Extractor",
//!   description: "Detects RDBMS (Diesel, SQLx, SeaORM), cache (Redis), document store (MongoDB), and message broker (Kafka, RabbitMQ, NATS) usage from derive macros, attributes, and source patterns.",
//!   since: "3.0.0"
//! }
//!
//! Looks for:
//! - Diesel ORM: `#[derive(Queryable)]`, `#[table_name]`, `diesel::table!`
//! - SQLx: `sqlx::query`, `#[sqlx(rename_all)]`, `Pool<Postgres>` / `Pool<Sqlite>`
//! - SeaORM: `#[derive(DeriveEntityModel)]`, `#[sea_orm(table_name)]`
//! - Redis: `redis` crate, `deadpool-redis`, `bb8-redis`
//! - Message brokers: `rdkafka`, `lapin` (RabbitMQ), `async-nats`

use serde_json::Value;
use quote::ToTokens;
use crate::scanner::FileInfo;
use super::{DocSpecExtractor, ProcessorContext};

/// Derive macros that indicate a database model.
const DB_DERIVES: &[&str] = &[
    "Queryable",
    "Insertable",
    "AsChangeset",
    "Identifiable",
    "Associations",
    "Selectable",
    "DeriveEntityModel",   // SeaORM
    "FromRow",             // SQLx
];

/// Attribute names that carry table name information.
const TABLE_ATTRS: &[&str] = &[
    "table_name",
    "diesel",
    "sea_orm",
    "sqlx",
];

/// @docspec:boundary "Data store detection across RDBMS, cache, document store, and message broker categories"
pub struct DataStoreExtractor;

impl DocSpecExtractor for DataStoreExtractor {
    /// @docspec:intentional "Check if any data store crate (diesel, sqlx, redis, kafka, etc.) is in Cargo.toml"
    fn is_available(&self, ctx: &ProcessorContext) -> bool {
        ctx.has_dependency("diesel")
            || ctx.has_dependency("sqlx")
            || ctx.has_dependency("sea-orm")
            || ctx.has_dependency("redis")
            || ctx.has_dependency("deadpool-redis")
            || ctx.has_dependency("rdkafka")
            || ctx.has_dependency("lapin")
            || ctx.has_dependency("async-nats")
            || ctx.has_dependency("rusqlite")
            || ctx.has_dependency("tokio-postgres")
            || ctx.has_dependency("mongodb")
    }

    /// @docspec:deterministic
    fn extractor_name(&self) -> &'static str {
        "data-store"
    }

    /// @docspec:intentional "Detect RDBMS tables, Redis, Kafka topics, RabbitMQ, NATS, and MongoDB from source patterns"
    fn extract(&self, files: &[FileInfo], ctx: &mut ProcessorContext) {
        let mut rdbms_tables: Vec<String> = Vec::new();
        let mut has_redis = false;
        let mut has_kafka = false;
        let mut kafka_topics: Vec<String> = Vec::new();
        let mut has_rabbitmq = false;
        let mut has_nats = false;
        let mut has_mongodb = false;
        let mut rdbms_type = detect_rdbms_type(ctx);

        for file_info in files {
            let syntax = match syn::parse_file(&file_info.source) {
                Ok(s) => s,
                Err(_) => continue,
            };

            for item in &syntax.items {
                match item {
                    syn::Item::Struct(s) => {
                        // Check if struct has DB model derives
                        if has_db_derive(&s.attrs) {
                            let table = extract_table_name(&s.attrs)
                                .unwrap_or_else(|| snake_case(&s.ident.to_string()));
                            if !rdbms_tables.contains(&table) {
                                rdbms_tables.push(table);
                            }
                        }
                    }
                    // Detect diesel::table! macro invocations
                    syn::Item::Macro(m) => {
                        let path_str = m.mac.path.to_token_stream().to_string();
                        if path_str.contains("table") {
                            let tokens = m.mac.tokens.to_string();
                            // First identifier in `table! { users { ... } }` is the table name
                            if let Some(name) = tokens.split_whitespace().next() {
                                let clean = name.trim_matches(|c: char| !c.is_alphanumeric() && c != '_');
                                if !clean.is_empty() && !rdbms_tables.contains(&clean.to_string()) {
                                    rdbms_tables.push(clean.to_string());
                                }
                            }
                        }
                    }
                    _ => {}
                }

                // Scan all items for Redis/Kafka/RabbitMQ/NATS/MongoDB usage
                let source = quote::quote!(#item).to_string();
                if source.contains("RedisConnection")
                    || source.contains("redis::cmd")
                    || source.contains("redis::Client")
                    || source.contains("redis::aio")
                    || source.contains("ConnectionManager")
                {
                    has_redis = true;
                }
                if source.contains("rdkafka")
                    || source.contains("StreamConsumer")
                    || source.contains("FutureProducer")
                    || source.contains("BaseProducer")
                {
                    has_kafka = true;
                    // Try to extract topic names from string literals near produce/subscribe calls
                    extract_kafka_topics(&source, &mut kafka_topics);
                }
                if source.contains("lapin") || source.contains("Channel") && source.contains("queue_declare") {
                    has_rabbitmq = true;
                }
                if source.contains("async_nats") || source.contains("nats::") {
                    has_nats = true;
                }
                if source.contains("mongodb::") || source.contains("Collection") && source.contains("Database") {
                    has_mongodb = true;
                }
            }
        }

        // Build data store entries
        if !rdbms_tables.is_empty() {
            rdbms_tables.sort();
            rdbms_tables.dedup();
            ctx.data_stores.push(serde_json::json!({
                "id": "rdbms",
                "name": rdbms_type,
                "type": "rdbms",
                "tables": rdbms_tables,
            }));
        }

        if has_redis {
            ctx.data_stores.push(serde_json::json!({
                "id": "redis",
                "name": "Redis",
                "type": "cache",
            }));
        }

        if has_kafka {
            kafka_topics.sort();
            kafka_topics.dedup();
            let mut store = serde_json::json!({
                "id": "kafka",
                "name": "Kafka",
                "type": "message-broker",
            });
            if !kafka_topics.is_empty() {
                let topics: Vec<Value> = kafka_topics.into_iter()
                    .map(|t| serde_json::json!({"name": t}))
                    .collect();
                store.as_object_mut().unwrap()
                    .insert("topics".to_string(), Value::Array(topics));
            }
            ctx.data_stores.push(store);
        }

        if has_rabbitmq {
            ctx.data_stores.push(serde_json::json!({
                "id": "rabbitmq",
                "name": "RabbitMQ",
                "type": "message-broker",
            }));
        }

        if has_nats {
            ctx.data_stores.push(serde_json::json!({
                "id": "nats",
                "name": "NATS",
                "type": "message-broker",
            }));
        }

        if has_mongodb {
            ctx.data_stores.push(serde_json::json!({
                "id": "mongodb",
                "name": "MongoDB",
                "type": "document-store",
            }));
        }
    }
}

/// Check if a struct has any database model derive macro.
///
/// @docspec:deterministic
fn has_db_derive(attrs: &[syn::Attribute]) -> bool {
    for attr in attrs {
        if attr.path().is_ident("derive") {
            let tokens = attr.meta.to_token_stream().to_string();
            for derive in DB_DERIVES {
                if tokens.contains(derive) {
                    return true;
                }
            }
        }
    }
    false
}

/// Extract a table name from `#[table_name = "users"]`, `#[diesel(table_name = users)]`,
/// or `#[sea_orm(table_name = "users")]`.
///
/// @docspec:deterministic
fn extract_table_name(attrs: &[syn::Attribute]) -> Option<String> {
    for attr in attrs {
        if let Some(ident) = attr.path().get_ident() {
            let name = ident.to_string();
            if TABLE_ATTRS.contains(&name.as_str()) {
                let tokens = attr.meta.to_token_stream().to_string();
                // Look for table_name = "..." or table_name = ident
                if let Some(pos) = tokens.find("table_name") {
                    let rest = &tokens[pos + 10..];
                    // Skip whitespace and = sign
                    let rest = rest.trim_start().trim_start_matches('=').trim_start();
                    if rest.starts_with('"') {
                        // Quoted string
                        if let Some(end) = rest[1..].find('"') {
                            return Some(rest[1..1 + end].to_string());
                        }
                    } else {
                        // Bare identifier
                        let ident_end = rest.find(|c: char| !c.is_alphanumeric() && c != '_')
                            .unwrap_or(rest.len());
                        let table = &rest[..ident_end];
                        if !table.is_empty() {
                            return Some(table.to_string());
                        }
                    }
                }
            }
        }
    }
    None
}

/// Convert CamelCase to snake_case for default table name derivation.
///
/// @docspec:deterministic
fn snake_case(s: &str) -> String {
    let mut result = String::new();
    for (i, c) in s.chars().enumerate() {
        if c.is_uppercase() && i > 0 {
            result.push('_');
        }
        result.push(c.to_lowercase().next().unwrap_or(c));
    }
    result
}

/// Detect the primary RDBMS type from Cargo.toml dependencies.
///
/// @docspec:deterministic
/// @docspec:intentional "Infers the RDBMS vendor (PostgreSQL, MySQL, SQLite) from Cargo.toml dependency names"
fn detect_rdbms_type(ctx: &ProcessorContext) -> String {
    if ctx.has_dependency("tokio-postgres") || ctx.cargo_toml.contains("Postgres") {
        "PostgreSQL".to_string()
    } else if ctx.cargo_toml.contains("Mysql") || ctx.cargo_toml.contains("mysql") {
        "MySQL".to_string()
    } else if ctx.has_dependency("rusqlite") || ctx.cargo_toml.contains("Sqlite") {
        "SQLite".to_string()
    } else {
        "Primary Database".to_string()
    }
}

/// Try to extract Kafka topic names from source text.
///
/// @docspec:intentional "Scan source text for topic/create_topic/subscribe keywords and extract quoted topic names"
fn extract_kafka_topics(source: &str, topics: &mut Vec<String>) {
    // Look for patterns like .create_topic("name", ...) or topic = "name"
    for keyword in &["topic", "create_topic", "subscribe"] {
        let mut search_from = 0;
        while let Some(pos) = source[search_from..].find(keyword) {
            let abs_pos = search_from + pos + keyword.len();
            if abs_pos < source.len() {
                let rest = &source[abs_pos..];
                // Find next quoted string
                if let Some(q_start) = rest.find('"') {
                    if let Some(q_end) = rest[q_start + 1..].find('"') {
                        let topic = &rest[q_start + 1..q_start + 1 + q_end];
                        if !topic.is_empty() && !topics.contains(&topic.to_string()) {
                            topics.push(topic.to_string());
                        }
                    }
                }
            }
            search_from = abs_pos;
        }
    }
}
